### PR TITLE
feat(http-client): Add parameters argument to get

### LINKED ISSUE #4: Feature request:  `params` argument for http-client.get()
Allow a `params` argument to be passed to the `get` method of aurelia-http-client. The get command would build a query string based on the object keys and values.

### PR DESCRIPTION
**Pull request by jedd-ahyoung**
Get request now accepts a parameters object and builds a query string based on its key/value pairs. Signature is now get(uri, params).

Fixes #4.

### CODE REVIEW

**Comment by santialbo** [comment]
Would be nice to have on all request types, not just get.

**Comment by jedd-ahyoung** [comment]
That can be done - I can add the `withParams` call to all of the http handlers. However, this will change the method signatures unless params is the last object for all of them.

**Comment by EisenbergEffect** [comment]
Let's hold off on this @jedd-ahyoung I think we need this...but we need to step back and think through the API design. Would you be willing to do some "prototype" api design around this? We can do it here in the comments...just throwing api's back and forth until we come up with something that will work more generally. Thoughts.

File: src/http-request-message.js [hunk 1]
```diff
@@ -16,8 +17,13 @@ export class HttpRequestMessage {
     return this;
   }
 
-  configureXHR(xhr){
-    xhr.open(this.method, this.uri, true);
+  withParams(params){
```
Comment (meenie) [inline src/http-request-message.js]: Why not use the ES6 default param here?

**Comment by jedd-ahyoung** [comment]
@meenie That's a good point. I overlooked that, still getting used to es6 features.

@EisenbergEffect That's a good idea.

In terms of the API, I suggested that we add a parameters object to the GET request. This, in turn, calls `withParams` on the actual `HTTPRequest` object; as such, `withParams` can be used manually on any request if so desired.

The `params` argument could easily be added to the existing methods. However, this would change the method signature, and, unless the argument were appended to the signature, would result in breaking changes (which may not be a huge deal now as we're still in preliminary).

If I were to change the signature, I'd suggest placing params second in the argument, so that all calls were uniform, like this:

    get(uri, params)
    post(uri, params, content, ...)
    ...

As for the HTTPRequest functionality, what do you guys think about the `withParams` call? Would it be better to specify that as a getter/setter on the HTTPRequest object?

Another option is to add querystring URL handling to `aurelia-path`. This might not be a bad place to put something like this, as it's really more of a URL utility tool than anything. However, anyone who wanted to use this functionality would have to pull in `aurelia-path` as a dependency, and it's not currently part of the Aurelia bundle.

All thoughts welcome!

**Comment by EisenbergEffect** [comment]
I'm wondering if we should try to expose a more fluent api in general....because there could be a lot of things people want to configure. What if we did something like this:

```javascipt
http.get(url)
  .withParams(params)
  .withHeader(key, value)
  .send();
```

It makes the simple use case without any options a bit uglier because you have to do this:

```javascript
http.get(url).send();
```

But...maybe we can have have a different api for one or the other...two versions? Or maybe a request buid (stream of consciousness here guys...):

```javascript
http.request.get(url)
  .withParams(params)
  .withHeader(key, value)
  .send();
```

Then leave the other apis alone. So you can use the basic verb apis for simple stuff and then use this request builder api for more complex stuff. Any thoughts on that?

**Comment by jedd-ahyoung** [comment]
Hmm....it seems a bit complex, but I think that the general idea isn't bad.

What about this? `http.request()` could return a wrapped HTTPRequest object. So you could do:

    http.request()
        .withParams(params)
        .withHeaders(headers)
        .get(url);

The nice thing about this is that it would still chain with promises without having to call `.send()`. Additionally, the call ends with the HTTP verb, which I personally like as it signifies the action (which as a general convention, I think would be more intuitive to JS developers). Finally, if you didn't want to customize the request, you could simply call `http.get(url)`; this effectively gives you a long-form call and a short-form call.

We could still provide a queryString builder in `aurelia-path`, for those who wouldn't want to build every request like this. Something like:

    import {addParams} from 'aurelia-path';

Then you could do something like `http.get(addParams(url, params))` without having to make a long call every time, just as another option. I'd shoot for a better name than 'addParams', though.

What do you think?

**Comment by jedd-ahyoung** [comment]
A final thought - perhaps it would be nice to eventually expose an API to decorate the request object. This would allow you do to do stuff like this:

    http.request()
        .withJWT(myJWT) // a custom function, added by a plugin perhaps?
        ...
        .post(url);

**Comment by meenie** [comment]
I'm liking @jedd-ahyoung's idea.  It allows for a lot of configurability by third parties.  Also makes it easy to write tests for.

**Comment by EisenbergEffect** [comment]
Ok. I like the direction this is going. Let's keep brainstorming here. Definitely the raw QS construction code can go in the path library. I think that makes sense.

What we need to nail down is how the request builder will work and how that could be extended by 3rd parties even. @jedd-ahyoung Would you be willing to fork this repo and go do some prototyping on that?

**Comment by EisenbergEffect** [comment]
@jedd-ahyoung I'm closing this PR for now. Can you open a new issue. We can continue design there and make plans to improve the http module.

**Comment by jedd-ahyoung** [comment]
No problem. I'll do a bit of prototyping tonight.

**Pull request closed without merging**